docs/man-openshmem/man1/oshmem_info.1.rst
.. _oshmem_info:


oshmem_info
===========

.. include_body

oshmem_info - Display information about the OpenSHMEM installation


SYNOPSIS
--------

``oshmem_info [options]``


DESCRIPTION
-----------

``oshmem_info`` provides detailed information about the OpenSHMEM
installation. It can be useful for at least three common scenarios:

1. Checking local configuration and seeing how OpenSHMEM was
   installed.

2. Submitting bug reports / help requests to the OpenSHMEM
   community
   (see :doc:`Getting help </getting-help>`).

3. Seeing a list of installed OpenSHMEM plugins and querying what MCA
   parameters they support.

.. note:: ``oshmem_info`` defaults to only showing a few MCA parameters
          by default (i.e., level 1 parameters). Use the ``--level``
          option to enable showing more options (see the :ref:`LEVELS
          <man1-oshmem_info-levels>` section for more information).


OPTIONS
-------

``oshmem_info`` accepts the following options:

* ``-a``, ``--all``: Show all configuration options and MCA parameters

* ``--arch``: Show architecture OpenSHMEM was compiled on

* ``-c``, ``--config``: Show configuration options

* ``-gmca``, ``--gmca <param> <value>*``: Pass global MCA parameters
  that are applicable to all contexts.

* ``-h``, ``--help*``: Shows help / usage message

* ``--hostname``: Show the hostname that OpenSHMEM was configured
  and built on

* ``--internal``: Show internal MCA parameters (not meant to be
  modified by users)

* ``--level``: Show only variables with at most this level (1-9). The
  default is 1 unless ``--all`` is specified without ``--level`` in
  which case the default is 9. See the :ref:`LEVELS
  <man1-oshmem_info-levels>` section for more information.

* ``-mca``, ``--mca <param> <value>``: Pass context-specific MCA
  parameters; they are considered global if
  ``--gmca`` is not used and only one context is specified.

* ``--param <type> <component>``: Show MCA parameters. The first
  parameter is the type of the component to display; the second
  parameter is the specific component to display (or the keyword
  ``all``, meaning "display all components of this type").

* ``--parsable``: When used in conjunction with other parameters,
  the output is displayed in a machine-parsable format ``--parseable``
  Synonym for ``--parsable``.

* ``--path <type>``: Show paths that OpenSHMEM was configured
  with. Accepts the following parameters: ``prefix``, ``bindir``,
  ``libdir``, ``incdir``, ``pkglibdir``, ``sysconfdir``.

* ``--pretty``: When used in conjunction with other parameters, the
  output is displayed in prettyprint format (default)

* ``--selected-only``: Show only variables from selected components.

* ``-V``, ``--version*``: Show version of OpenSHMEM.

.. _man1-oshmem_info-levels:

LEVELS
------

#. Basic information of interest to users
#. Detailed information of interest to users
#. All remaining information of interest to users
#. Basic information required for tuning
#. Detailed information required for tuning
#. All remaining information required for tuning
#. Basic information for OpenSHMEM implementors
#. Detailed information for OpenSHMEM implementors
#. All remaining information for OpenSHMEM implementors


EXAMPLES
--------

.. code:: sh

   shell$ oshmem_info

Show the default output of options and listing of installed components
in a human-readable / prettyprint format.

.. code:: sh

   shell$ oshmem_info --parsable

Show the default output of options and listing of installed components
in a machine-parsable format.

.. code:: sh

   shell$ oshmem_info --param btl openib

Show the MCA parameters of the "openib" BTL component in a
human-readable / prettyprint format.

.. code:: sh

   shell$ oshmem_info --param btl openib --parsable

Show the MCA parameters of the "openib" BTL component in a
machine-parsable format.

.. code:: sh

   shell$ oshmem_info --path bindir

Show the "bindir" that OpenSHMEM was configured with.

.. code:: sh

   shell$ oshmem_info --version ompi full --parsable

Show the full version numbers of OpenSHMEM (including the ORTE
and OPAL version numbers) in a machine-readable format.

.. code:: sh

   shell$ oshmem_info --version btl major

Show the major version number of all BTL components in a prettyprint
format.

.. code:: sh

   shell$ oshmem_info --version btl:tcp minor

Show the minor version number of the TCP BTL component in a
prettyprint format.

.. code:: sh

   shell$ oshmem_info --all

Show *all* information about the OpenSHMEM installation, including all
components that can be found, the MCA parameters that they support,
versions of OpenSHMEM and the components, etc.


NEWS
Copyright (c) 2004-2010 The Trustees of Indiana University and Indiana
                        University Research and Technology
                        Corporation.  All rights reserved.
Copyright (c) 2004-2006 The University of Tennessee and The University
                        of Tennessee Research Foundation.  All rights
                        reserved.
Copyright (c) 2004-2008 High Performance Computing Center Stuttgart,
                        University of Stuttgart.  All rights reserved.
Copyright (c) 2004-2006 The Regents of the University of California.
                        All rights reserved.
Copyright (c) 2006-2022 Cisco Systems, Inc.  All rights reserved.
Copyright (c) 2006      Voltaire, Inc. All rights reserved.
Copyright (c) 2006      Sun Microsystems, Inc.  All rights reserved.
                        Use is subject to license terms.
Copyright (c) 2006-2018 Los Alamos National Security, LLC.  All rights
                        reserved.
Copyright (c) 2010-2020 IBM Corporation.  All rights reserved.
Copyright (c) 2012      Oak Ridge National Labs.  All rights reserved.
Copyright (c) 2012      Sandia National Laboratories.  All rights reserved.
Copyright (c) 2012      University of Houston. All rights reserved.
Copyright (c) 2013      NVIDIA Corporation.  All rights reserved.
Copyright (c) 2013-2018 Intel, Inc. All rights reserved.
Copyright (c) 2018-2021 Amazon.com, Inc. or its affiliates.  All Rights
                        reserved.
Copyright (c) 2019-2021 Triad National Security, LLC. All rights
                        reserved.
Copyright (c) 2021      Google, LLC. All rights reserved.
$COPYRIGHT$

Additional copyrights may follow

$HEADER$

===========================================================================

This file contains the main features as well as overviews of specific
bug fixes (and other actions) for each version of Open MPI since
version 1.0.

As more fully described in the "Software Version Number" section in
the README file, Open MPI typically releases two separate version
series simultaneously.  Since these series have different goals and
are semi-independent of each other, a single NEWS-worthy item may be
introduced into different series at different times.  For example,
feature F was introduced in the vA.B series at version vA.B.C, and was
later introduced into the vX.Y series at vX.Y.Z.

The first time feature F is released, the item will be listed in the
vA.B.C section, denoted as:

   (** also to appear: X.Y.Z) -- indicating that this item is also
                                 likely to be included in future release
                                 version vX.Y.Z.

When vX.Y.Z is later released, the same NEWS-worthy item will also be
included in the vX.Y.Z section and be denoted as:

   (** also appeared: A.B.C)  -- indicating that this item was previously
                                 included in release version vA.B.C.

4.1.3 -- March, 2022
--------------------

- Fixed a seg fault in the smcuda BTL.  Thanks to Moritz Kreutzer and
  @Stadik for reporting the issue.
- Added support for ELEMENTAL to the MPI handle comparison functions
  in the mpi_f08 module.  Thanks to Salvatore Filippone for raising
  the issue.
- Minor datatype performance improvements in the CUDA-based code paths.
- Fix MPI_ALLTOALLV when used with MPI_IN_PLACE.
- Fix MPI_BOTTOM handling for non-blocking collectives.  Thanks to
  Lisandro Dalcin for reporting the problem.
- Enable OPAL memory hooks by default for UCX.
- Many compiler warnings fixes, particularly for newer versions of
  GCC.
- Fix intercommunicator overflow with large payload collectives.  Also
  fixed MPI_REDUCE_SCATTER_BLOCK for similar issues with large payload
  collectives.
- Back-port ROMIO 3.3 fix to use stat64() instead of stat() on GPFS.
- Fixed several non-blocking MPI collectives to not round fractions
  based on float precision.
- Fix compile failure for --enable-heterogeneous.  Also updated the
  README to clarify that --enable-heterogeneous is functional, but
  still not recomended for most environments.
- Minor fixes to OMPIO, including:
  - Fixing the open behavior of shared memory shared file pointers.
    Thanks to Axel Huebl for reporting the issue
  - Fixes to clean up lockfiles when closing files.  Thanks to Eric
    Chamberland for reporting the issue.
- Update LSF configure failure output to be more clear (e.g., on RHEL
  platforms).
- Update if_[in|ex]clude behavior in btl_tcp and oob_tcp to select
  *all* interfaces that fall within the specified subnet range.


4.1.2 -- November, 2021
-----------------------

- ROMIO portability fix for OpenBSD
- Fix handling of MPI_IN_PLACE with MPI_ALLTOALLW and improve performance
  of MPI_ALLTOALL and MPI_ALLTOALLV for MPI_IN_PLACE.
- Fix one-sided issue with empty groups in Post-Start-Wait-Complete
  synchronization mode.
- Fix Fortran status returns in certain use cases involving
  Generalized Requests
- Romio datatype bug fixes.
- Fix oshmem_shmem_finalize() when main() returns non-zero value.
- Fix wrong affinity under LSF with the membind option.
- Fix count==0 cases in MPI_REDUCE and MPI_IREDUCE.
- Fix ssh launching on Bourne-flavored shells when the user has "set
  -u" set in their shell startup files.
- Correctly process 0 slots with the mpirun --host option.
- Ensure to unlink and rebind socket when the Open MPI session
  directory already exists.
- Fix a segv in mpirun --disable-dissable-map.
- Fix a potential hang in the memory hook handling.
- Slight performance improvement in MPI_WAITALL when running in
  MPI_THREAD_MULTIPLE.
- Fix hcoll datatype mapping and rooted operation behavior.
- Correct some operations modifying MPI_Status.MPI_ERROR when it is
  disallowed by the MPI standard.
- UCX updates:
  - Fix datatype reference count issues.
  - Detach dynamic window memory when freeing a window.
  - Fix memory leak in datatype handling.
- Fix various atomic operations issues.

oshmem/tools/oshmem_info/oshmem_info.c
/*
 * Copyright (c) 2013-2015 Mellanox Technologies, Inc.
 *                         All rights reserved.
 * Copyright (c) 2014-2019 Intel, Inc.  All rights reserved.
 *
 * Copyright (c) 2016-2017 IBM Corporation. All rights reserved.
 * $COPYRIGHT$
 *
 * Additional copyrights may follow
 *
 * $HEADER$
 */

#include "oshmem_config.h"

#include <ctype.h>
#include <stdio.h>
#include <string.h>
#ifdef HAVE_UNISTD_H
#    include <unistd.h>
#endif
#ifdef HAVE_NETDB_H
#    include <netdb.h>
#endif
#ifdef HAVE_SYS_PARAM_H
#    include <sys/param.h>
#endif
#include <errno.h>
#include <signal.h>

#include "opal/class/opal_object.h"
#include "opal/class/opal_pointer_array.h"
#include "opal/mca/base/base.h"
#include "opal/mca/installdirs/installdirs.h"
#include "opal/runtime/opal.h"
#include "opal/runtime/opal_info_support.h"
#include "opal/util/argv.h"
#include "opal/util/show_help.h"
#include "opal/version.h"

#include "ompi/runtime/ompi_info_support.h"

#include "oshmem/include/oshmem/constants.h"
#include "oshmem/runtime/oshmem_info_support.h"
#include "oshmem/tools/oshmem_info/oshmem_info.h"

/*
 * Public variables
 */

int main(int argc, char *argv[])
{
    int ret = 0;
    bool acted = false;
    bool want_all = false;
    int i;
    opal_cmd_line_t *info_cmd_line;
    opal_pointer_array_t mca_types;
    opal_pointer_array_t component_map;
    opal_info_component_map_t *map;

    /* protect against problems if someone passes us thru a pipe
     * and then abnormally terminates the pipe early */
    signal(SIGPIPE, SIG_IGN);

    /* Initialize the argv parsing handle */
    if (OPAL_SUCCESS != opal_init_util(&argc, &argv)) {
        opal_show_help("help-oshmem-info.txt", "lib-call-fail", true, "opal_init_util", __FILE__,
                       __LINE__, NULL);
        exit(ret);
    }

    info_cmd_line = OBJ_NEW(opal_cmd_line_t);
    if (NULL == info_cmd_line) {
        ret = errno;
        opal_show_help("help-oshmem-info.txt", "lib-call-fail", true, "opal_cmd_line_create",
                       __FILE__, __LINE__, NULL);
        exit(ret);
    }

    /* initialize the command line, parse it, and return the directives
     * telling us what the user wants output
     */
    if (OPAL_SUCCESS != (ret = opal_info_init(argc, argv, info_cmd_line))) {
        exit(ret);
    }

    if (opal_cmd_line_is_taken(info_cmd_line, "version")) {
        fprintf(stdout, "Open MPI/SHMEM v%s\n\n%s\n", OPAL_VERSION, PACKAGE_BUGREPORT);
        exit(0);
    }

    /* setup the mca_types array */
    OBJ_CONSTRUCT(&mca_types, opal_pointer_array_t);
    opal_pointer_array_init(&mca_types, 256, INT_MAX, 128);

    /* add in the opal frameworks */
    opal_info_register_types(&mca_types);

    /* add in the ompi frameworks */
    opal_info_register_types(&mca_types);

    /* add in the oshmem frameworks */
    oshmem_info_register_types(&mca_types);

    /* init the component map */
    OBJ_CONSTRUCT(&component_map, opal_pointer_array_t);
    opal_pointer_array_init(&component_map, 256, INT_MAX, 128);

    /* Register OMPI/OSHMEM's params */
    if (OSHMEM_SUCCESS != (ret = oshmem_info_register_framework_params(&component_map))) {
        if (OSHMEM_ERR_BAD_PARAM == ret) {
            /* output what we got */
            opal_info_do_params(true, opal_cmd_line_is_taken(info_cmd_line, "internal"), &mca_types,
                                &component_map, NULL);
        }
        exit(1);
    }

    /* Execute the desired action(s) */
    want_all = opal_cmd_line_is_taken(info_cmd_line, "all");
    if (want_all || opal_cmd_line_is_taken(info_cmd_line, "path")) {
        opal_info_do_path(want_all, info_cmd_line);
        acted = true;
    }
    if (want_all || opal_cmd_line_is_taken(info_cmd_line, "arch")) {
        opal_info_do_arch();
        acted = true;
    }
    if (want_all || opal_cmd_line_is_taken(info_cmd_line, "hostname")) {
        opal_info_do_hostname();
        acted = true;
    }
    if (want_all || opal_cmd_line_is_taken(info_cmd_line, "config")) {
        oshmem_info_do_config(true);
        acted = true;
    }
    if (want_all || opal_cmd_line_is_taken(info_cmd_line, "param")
        || opal_cmd_line_is_taken(info_cmd_line, "params")) {
        opal_info_do_params(want_all, opal_cmd_line_is_taken(info_cmd_line, "internal"), &mca_types,
                            &component_map, info_cmd_line);
        acted = true;
    }

    /* If no command line args are specified, show default set */

    if (!acted) {
        opal_info_out("Package", "package", OPAL_PACKAGE_STRING);
        oshmem_info_show_oshmem_version(opal_info_ver_full);
        opal_info_show_path(opal_info_path_prefix, opal_install_dirs.prefix);
        opal_info_do_arch();
        opal_info_do_hostname();
        oshmem_info_do_config(false);
        opal_info_show_component_version(&mca_types, &component_map, opal_info_type_all,
                                         opal_info_component_all, opal_info_ver_full,
                                         opal_info_ver_all);
    }

    /* All done */

    oshmem_info_close_components();
    OBJ_RELEASE(info_cmd_line);
    OBJ_DESTRUCT(&mca_types);
    for (i = 0; i < component_map.size; i++) {
        if (NULL
            != (map = (opal_info_component_map_t *) opal_pointer_array_get_item(&component_map,
                                                                                i))) {
            OBJ_RELEASE(map);
        }
    }
    OBJ_DESTRUCT(&component_map);

    opal_info_finalize();

    /* Put our own call to opal_finalize_util() here because we called
       it up above (and it refcounts) */
    opal_finalize_util();

    return 0;
}
